README.txt
/@@localroles-controlpanel
**************************
**Info**
    requires Plone4

Add or Remove local roles for objects using csv. 
The format of the csv file should be::

    path,user-/group-id,role

The data gets processed line by line. 
At the end, you'll get a notification on the number of 
successfully processed lines as well as 
verbose error notifications for each unsuccessful line. 
**Be patient!** The duration of the process depends on 
the number of lines of your csv and may take up to 
several minutes.

Example csv::

        /Plone/front-page,Reviewers,Editor
        /Plone/front-page,usera,Contributor
        /Plone/news,userb,Reviewer
        /Plone/news,userb,Editor


docs/HISTORY.txt
Changelog
=========

0.2 (2012-03-07)
----------------

- made compatible with Plone 4.1.x
  [hplocher]

0.1 (2010-08-02)
----------------

- Initial release


collective/localrolespanel/tests/base.py
"""Test setup for integration and functional tests.

When we import PloneTestCase and then call setupPloneSite(), all of
Plone's products are loaded, and a Plone site will be created. This
happens at module level, which makes it faster to run each test, but
slows down test runner startup.
"""

from Products.Five import zcml
from Products.Five import fiveconfigure

from Testing import ZopeTestCase as ztc

from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import onsetup

# When ZopeTestCase configures Zope, it will *not* auto-load products
# in Products/. Instead, we have to use a statement such as:
#   ztc.installProduct('SimpleAttachment')
# This does *not* apply to products in eggs and Python packages (i.e.
# not in the Products.*) namespace. For that, see below.
# All of Plone's products are already set up by PloneTestCase.


@onsetup
def setup_product():
    """Set up the package and its dependencies.

    The @onsetup decorator causes the execution of this body to be
    deferred until the setup of the Plone site testing layer. We could
    have created our own layer, but this is the easiest way for Plone
    integration tests.
    """

    # Load the ZCML configuration for the example.tests package.
    # This can of course use <include /> to include other packages.

    fiveconfigure.debug_mode = True
    import collective.localrolespanel
    zcml.load_config('configure.zcml', collective.localrolespanel)
    fiveconfigure.debug_mode = False

    # We need to tell the testing framework that these products
    # should be available. This can't happen until after we have loaded
    # the ZCML. Thus, we do it here. Note the use of installPackage()
    # instead of installProduct().
    # This is *only* necessary for packages outside the Products.*
    # namespace which are also declared as Zope 2 products, using
    # <five:registerPackage /> in ZCML.

    # We may also need to load dependencies, e.g.:
    #   ztc.installPackage('borg.localrole')

    ztc.installPackage('collective.localrolespanel')

# The order here is important: We first call the (deferred) function
# which installs the products we need for this product. Then, we let
# PloneTestCase set up this product on installation.

setup_product()
ptc.setupPloneSite(products=['collective.localrolespanel'])


class TestCase(ptc.PloneTestCase):
    """We use this base class for all the tests in this package. If
    necessary, we can put common utility or setup code in here. This
    applies to unit test cases.
    """


class FunctionalTestCase(ptc.FunctionalTestCase):
    """We use this class for functional integration tests that use
    doctest syntax. Again, we can put basic common utility or setup
    code in here.
    """

    def afterSetUp(self):
        roles = ('Member', 'Contributor')
        self.portal.portal_membership.addMember('contributor',
                                                'secret',
                                                roles, [])


collective/localrolespanel/tests/test_setup.py
# -*- coding: utf-8 -*-
#
# File: test_setup.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = """Hans-Peter Locher <hans-peter.locher@inquant.de>"""
__docformat__ = 'plaintext'


import unittest

from zope import interface
from zope import component

from Products.CMFCore.utils import getToolByName

from collective.localrolespanel.tests.base import TestCase


from collective.localrolespanel.interfaces import IStdLocalRoles
from collective.localrolespanel.interfaces import StdLocalRolesError

class TestStdLocalRoles(TestCase):

    def afterSetUp(self):
        self.folder.invokeFactory("Document", "testcontext")
        self.testcontext = self.folder.testcontext
        self.portal.acl_users._doAddUser('testuser', 'secret', [], [])

    def test_utility_registered(self):
        rolesapi = component.getUtility(IStdLocalRoles)
        self.failUnless(IStdLocalRoles.providedBy(rolesapi))

    def test_addRole(self):
        rolesapi = component.getUtility(IStdLocalRoles)
        rolesapi.addRole(self.testcontext, 'testuser', 'Reader')
        self.assertEquals(self.testcontext.get_local_roles_for_userid(userid='testuser'), ('Reader',))
        rolesapi.addRole(self.testcontext, 'testuser', 'Member')
        self.assertEquals(self.testcontext.get_local_roles_for_userid(userid='testuser'), ('Reader', 'Member'))
        rolesapi.addRole(self.testcontext, 'testuser', 'Reader')
        self.assertEquals(self.testcontext.get_local_roles_for_userid(userid='testuser'), ('Reader', 'Member'))

    def test_addRole_unknown_principal(self):
        rolesapi = component.getUtility(IStdLocalRoles)
        self.failUnlessRaises(StdLocalRolesError, rolesapi.addRole, self.testcontext, 'unknownuser', 'Reader')

    def test_addRole_unknown_role(self):
        rolesapi = component.getUtility(IStdLocalRoles)
        self.failUnlessRaises(StdLocalRolesError, rolesapi.addRole, self.testcontext, 'testuser', 'unknownrole')

    def test_removeRole(self):
        rolesapi = component.getUtility(IStdLocalRoles)
        rolesapi.addRole(self.testcontext, 'testuser', 'Reader')
        self.assertEquals(self.testcontext.get_local_roles_for_userid(userid='testuser'), ('Reader',))
        rolesapi.addRole(self.testcontext, 'testuser', 'Member')
        self.assertEquals(self.testcontext.get_local_roles_for_userid(userid='testuser'), ('Reader', 'Member'))
        rolesapi.removeRole(self.testcontext, 'testuser', 'Reader')
        self.assertEquals(self.testcontext.get_local_roles_for_userid(userid='testuser'), ('Member',))
        rolesapi.removeRole(self.testcontext, 'testuser', 'Member')
        self.assertEquals(self.testcontext.get_local_roles_for_userid(userid='testuser'), ())

def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(TestStdLocalRoles))
    return suite

# vim: set ft=python ts=4 sw=4 expandtab :


collective/localrolespanel/tests/test_doctest.py
import unittest
import doctest

#from zope.testing import doctestunit
#from zope.component import testing, eventtesting

from Testing import ZopeTestCase as ztc

from collective.localrolespanel.tests import base


def test_suite():
    return unittest.TestSuite([

        # Demonstrate the main content types
        ztc.ZopeDocFileSuite(
            'README.txt', package='collective.localrolespanel',
            test_class=base.FunctionalTestCase,
            optionflags=doctest.REPORT_ONLY_FIRST_FAILURE |
                doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS),

        ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


collective/localrolespanel/browser/localroles_controlpanel.py
# -*- coding: utf-8 -*-
#
# File: localroles_controlpanel.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = 'Ramon Bartl <ramon.bartl@inquant.de>'
__docformat__ = 'plaintext'

import csv
import logging
from StringIO import StringIO

from zope import schema
from zope import component
from zope import interface
from zope.formlib import form as formlib

from plone.protect import CheckAuthenticator
from plone.app.form.validators import null_validator
from five.formlib.formbase import PageForm
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile

from Products.statusmessages.interfaces import IStatusMessage
from Products.CMFCore.utils import getToolByName

from collective.localrolespanel.interfaces import IStdLocalRoles
from collective.localrolespanel.interfaces import StdLocalRolesError

from collective.localrolespanel import localrolespanelMessageFactory as _

DELIMITER = ","
logger = logging.getLogger("localroles_controlpanel")


class ILocalRoleSettings(interface.Interface):
    """ Local Roles Bulk Settings
    """

    csv_file = schema.Bytes(title=_(u'label_csv_file',
                                    default=_(u'Select CSV file')),
                        description=_(u'help_csv_file',
                                      default=u'Select a CSV File'),
                        required=True)


class LocalRolesControlPanel(PageForm):
    """ Local roles control panel for bulk settings
    """

    template = ViewPageTemplateFile('localroles_controlpanel_form.pt')

    def __init__(self, context, request):
        PageForm.__init__(self, context, request)
        self.utility = component.getUtility(IStdLocalRoles)


    def validate_csv(self, action, data):
        """ csv validation
        """
        # CSRF protection
        CheckAuthenticator(self.request)


    @property
    def form_fields(self):
        """ Form Fields
            see: ILocalRoleSettings
        """
        return formlib.Fields(ILocalRoleSettings)


    @formlib.action(_(u'label_add_roles', default=u'Add Roles'),
                 validator='validate_csv', name=u'Add Roles')
    def action_add_roles(self, action, data):
        """ handle add roles
        """
        errors = []
        success = 0

        csv_file = StringIO(data["csv_file"])
        reader = csv.reader(csv_file, delimiter=DELIMITER)

        for num, row in enumerate(reader):
            if len(row) != 3:
                errors.append("Line %(num)s, %(row)s, Invalid line" % locals())
                continue

            try:
                content_item = self.context.unrestrictedTraverse(row[0])
            except KeyError, e:
                errors.append("Line %(num)s, %(row)s, Object not found" % locals())
                continue

            try:
                self.utility.addRole(content_item, row[1], row[2])
                success += 1
            except StdLocalRolesError, e:
                errors.append("Line %(num)s, %(row)s, %(e)s" % locals())

        msg = "Successfully added Roles for %s Lines" % str(success)
        if errors:
            self._add_status_message(msg)

            for e in errors:
                self._add_status_message(str(e), severity="error")
        else:
            self._add_status_message(msg)

        self._redirect_back_to(url="/@@localroles-controlpanel")


    @formlib.action(_(u'label_remove_roles', default=u'Remove Roles'),
                 validator='validate_csv', name=u'Remove Roles')
    def action_remove_roles(self, action, data):
        """ handle remove roles
        """
        errors = []
        success = 0

        csv_file = StringIO(data["csv_file"])
        reader = csv.reader(csv_file, delimiter=DELIMITER)

        for num, row in enumerate(reader):
            if len(row) != 3:
                errors.append("Line %(num)s, %(row)s, Invalid line" % locals())
                continue

            try:
                content_item = self.context.unrestrictedTraverse(row[0])
            except KeyError, e:
                errors.append("Line %(num)s, %(row)s, Object not found" % locals())
                continue

            try:
                self.utility.removeRole(content_item, row[1], row[2])
                success += 1
            except StdLocalRolesError, e:
                errors.append("Line %(num)s, %(row)s, %(e)s" % locals())

        msg = "Successfully removed Roles for %s Lines" % str(success)

        if errors:
            self._add_status_message(msg)
            for e in errors:
                self._add_status_message(str(e), severity="error")
        else:
            self._add_status_message(msg)

        self._redirect_back_to(url="/@@localroles-controlpanel")


    @formlib.action(_(u'label_cancel', default=u'Cancel'),
                 validator=null_validator,
                 name=u'cancel')
    def handle_cancel_action(self, action, data):
        """ Cancel Action
            -> redirects to plone_control_panel
        """
        self._add_status_message("No Changes made")
        self._redirect_back_to(url="/plone_control_panel")


    def _add_status_message(self, msg, severity="info"):
        """ append a status message
        """
        IStatusMessage(self.request).addStatusMessage(msg, severity)


    def _redirect_back_to(self, url=""):
        """ redirect to url
        """
        context_url = component.getMultiAdapter((self.context, self.request),
                name='absolute_url')()
        self.request.response.redirect(context_url + url)

# vim: set ft=python ts=4 sw=4 expandtab :


collective/localrolespanel/stdlocalroles.py
# -*- coding: utf-8 -*-
#
# File: stdlocalroles.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = """Hans-Peter Locher <hans-peter.locher@inquant.de>"""
__docformat__ = 'plaintext'


import logging

from zope import interface
from zope.app.component.hooks import getSite

from interfaces.stdlocalroles import IStdLocalRoles
from interfaces.stdlocalroles import StdLocalRolesError
from Products.CMFCore.utils import getToolByName

LOGGER = logging.getLogger("collective.localrolespanel")

class StdLocalRoles(object):
    """ API for Plone std local roles
    """
    interface.implements(IStdLocalRoles)

    def validate_parameters(self, context, principal_id, role):
        """validates given parameters, raises an StdLocalRolesError
           with msg if not valid, otherwise returns True
        """
        groups = getToolByName(context, "portal_groups")
        users = getToolByName(context, "portal_membership")
        valid_principal_id = groups.getGroupById(principal_id) or users.getMemberById(principal_id)
        if not valid_principal_id:
            msg = "Unknown user/group id: %s" % principal_id
            raise StdLocalRolesError, msg
        valid_roles = context.valid_roles()
        if not role in valid_roles:
            msg = "Unknown role: %s, valid roles are: %s" % (role, valid_roles)
            raise StdLocalRolesError, msg
        return True

    def addRole(self, context, principal_id, role):
        """Adds the specified role to the context for principal_id
           returns True or raises an StdLocalRolesError
        """
        valid = self.validate_parameters(context, principal_id, role)
        if role not in context.get_local_roles_for_userid(userid=principal_id):
            context.manage_addLocalRoles(principal_id, [role])
            context.reindexObjectSecurity()
        return True

    def removeRole(self, context, principal_id, role):
        """Removes the specified role from the context for principal_id
           returns True or raises an StdLocalRolesError
        """
        valid = self.validate_parameters(context, principal_id, role)
        existing = context.get_local_roles_for_userid(userid=principal_id)
        if role in existing:
            kept_roles = [kept_role for kept_role in existing if kept_role != role]
            context.manage_delLocalRoles(userids=[principal_id])
            if len(kept_roles):
                context.manage_addLocalRoles(principal_id, kept_roles)
            context.reindexObjectSecurity()
        return True


# vim: set ft=python ts=4 sw=4 expandtab :
